cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -345,6 +362,138 @@ public ApplicationWithPrograms updateApp(ApplicationId appId, AppRequest appRequ
+    if (currentArtifact.getVersion().compareTo(candidateArtifactVersion) > 0) {
+      String error = String.format(
+        "The current artifact has a version higher %s than any existing artifact.", currentArtifact.getVersion());
+      return new ApplicationUpdateDetail(appId, "Upgrade failed: " + error);
+    }
+
+    ArtifactId newArtifactId =
+      new ArtifactId(currentArtifact.getName(), candidateArtifactVersion, currentArtifact.getScope());
+
+    Id.Artifact newArtifact = Id.Artifact.fromEntityId(Artifacts.toProtoArtifactId(appId.getParent(), newArtifactId));
+    ArtifactDetail newArtifactDetail = artifactRepository.getArtifact(newArtifact);
+    List<ApplicationConfigUpdateAction> upgradeActions = Arrays.asList(ApplicationConfigUpdateAction.UPGRADE_ARTIFACT);
+
+
+    return updateApplicationInternal(appId, currentSpec.getConfiguration(), programTerminator, newArtifactDetail,
+                                     upgradeActions, ownerAdmin.getOwner(appId), false, "Upgrade");
+  }
+
+  /**
+   * Updates an application config by applying given update actions. The app should know how to apply these actions
+   * to its config.
+   */
+  private ApplicationUpdateDetail updateApplicationInternal(ApplicationId appId,
+                                                            @Nullable String currentConfigStr,
+                                                            ProgramTerminator programTerminator,
+                                                            ArtifactDetail artifactDetail,
+                                                            List<ApplicationConfigUpdateAction> updateActions,
+                                                            @Nullable KerberosPrincipalId ownerPrincipal,
+                                                            boolean updateSchedules,
+                                                            String userAction) throws Exception {
+    ApplicationClass appClass = Iterables.getFirst(artifactDetail.getMeta().getClasses().getApps(), null);
+    if (appClass == null) {
+      // This should never happen.
+      throw new IllegalStateException(String.format("No application class found in artifact '%s' in namespace '%s'.",
+                                      artifactDetail.getDescriptor().getArtifactId(), appId.getParent()));
+    }
+    io.cdap.cdap.proto.id.ArtifactId artifactId =
+      Artifacts.toProtoArtifactId(appId.getParent(), artifactDetail.getDescriptor().getArtifactId());
+    EntityImpersonator classLoaderImpersonator = new EntityImpersonator(artifactId, this.impersonator);
+
+    String updatedAppConfig = "";
+    DefaultApplicationUpdateContext updateContext =
+      new DefaultApplicationUpdateContext(appId.getParent(), appId, artifactDetail.getDescriptor().getArtifactId(),
+                                          artifactRepository, currentConfigStr, updateActions);
+
+    Object appMain;
+    try {
+      ClassLoader artifactClassLoader =
```

Comment:
You need to close this classloader after done using any instances created from it. The createArtifactClassLoader returns a `CloseableClassLoader`, which you can use `try (resource)` block to easily close it out.